src/easylink/step.py

```diff
@@ -249,39 +91,236 @@ def __init__(
+
+        This compares the ``Step`` *instance* name to its *node* name via the ``Step's``
+        ordered hierarchy of sub-``Steps`` and uses the full suffix of names starting
+        from wherever the two first differ.
+
+        For example, a ``Step`` named "step_3" may loop multiple times using the same
+        :class:`~easylink.implementation.Implementation` named "step_3_python_pandas".
+        However, to disambiguate between the different loops of "step_3", we might
+        designate the node name to be "step_3_loop_1" and then combine that with the
+        ``Implementation`` name such that the ``Implementation's`` node name is
+        "step_3_loop_1_step_3_python_pandas".
+
+        If all the node names and step names match, we have not introduced any step
+        degeneracies (with e.g. loops or multiples), and we can simply use the
+        implementation name directly.
+
+        Returns
+        -------
+            The unique name to be used for this ``Step's`` node in the ``ImplementationGraph``.
+        """
+        step = self
+        implementation_name = (
+            self.configuration_state.pipeline_config[COMBINED_IMPLEMENTATION_KEY]
+            if self.configuration_state.is_combined
+            else self.configuration_state.implementation_config.name
+        )
+        node_names = []
+        step_names = []
+        while step:
+            node_names.append(step.name)
+            step_names.append(step.step_name)
+            step = step.parent_step
+
+        prefix = []
+        step_names.reverse()
+        node_names.reverse()
+        for i, (step_name, node_name) in enumerate(zip(step_names, node_names)):
+            if step_name != node_name:
+                prefix = node_names[i:]
+                break
+        # If we didn't include the step name already for a combined implementation, do so now.
+        if self.configuration_state.is_combined and not prefix:
+            prefix.append(self.name)
+        prefix.append(implementation_name)
+        return "_".join(prefix)
+
+    ###########
+    # Methods #
+    ###########
+
+    def validate_step(
         self,
         step_config: LayeredConfigTree,
         combined_implementations: LayeredConfigTree,
+        input_data_config: LayeredConfigTree,
     ) -> dict[str, list[str]]:
+        """Validates the ``Step``.
+
+        Parameters
+        ----------
+        step_config
+            The configuration of this ``Step``.
+        combined_implementations
+            The relevant configuration if this ``Step's`` :class:`~easylink.implementation.Implementation`
+            has been requested to be combined with that of a different ``Step``.
+        input_data_config
+            The input data configuration for the entire pipeline.
+
+        Returns
+        -------
+            A dictionary of errors, where the keys are the ``Step`` name and the
+            values are lists of as many error messages as could be generated.
```

Comment:
```suggestion
            values are lists of error messages associated with the given step.
```

And then add a `Note` below saying "We make every attempt to batch error messages,
but there are cases where the configuration is so ill-formed that fixing one error will
'reveal' another the next time the configuration is validated."